src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilationModuleGroupBase.cs

```diff
@@ -0,0 +1,293 @@
+
+                foreach (var module in _compilationModuleSet)
+                {
+                    EcmaModule ecmaModule = (EcmaModule)module;
+                    foreach (var typeRefHandle in ecmaModule.MetadataReader.TypeReferences)
+                    {
+                        try
+                        {
+                            TypeDesc typeFromTypeRef = ecmaModule.GetType(typeRefHandle);
+                            if (!_typeRefsInCompilationModuleSet.ContainsKey(typeFromTypeRef))
+                            {
+                                _typeRefsInCompilationModuleSet.Add(typeFromTypeRef, new ModuleToken(ecmaModule, typeRefHandle));
+                            }
+                        }
+                        catch (TypeSystemException) { }
+                    }
+                }
+            }
+
+            return _typeRefsInCompilationModuleSet.TryGetValue(type, out token);
+        }
+
+        private bool ComputeTypeVersionsWithCode(TypeDesc type)
+        {
+            if (type == type.Context.CanonType)
+                return true;
+
+            if (type is MetadataType mdType)
+            {
+                if (!_versionBubbleModuleSet.Contains(mdType.Module))
+                    return false;
+            }
+
+            if (type == type.GetTypeDefinition())
+                return true;
+
+            return ComputeInstantiationVersionsWithCode(type.Instantiation, type);
+        }
+
+        private bool ComputeInstantiationVersionsWithCode(Instantiation inst, TypeSystemEntity entityWithInstantiation)
+        {
+            for (int iInstantiation = 0; iInstantiation < inst.Length; iInstantiation++)
+            {
+                TypeDesc instType = inst[iInstantiation];
+
+                if (!ComputeInstantiationTypeVersionsWithCode(instType))
+                {
+                    if (instType.IsPrimitive)
+                    {
+                        // Primitive type instantiations are only instantiated in the module of the generic defining type
+                        // if the generic does not apply interface constraints to that type parameter, or if System.Private.CoreLib is part of the version bubble
+
+                        Instantiation entityDefinitionInstantiation;
+                        if (entityWithInstantiation is TypeDesc type)
+                        {
+                            entityDefinitionInstantiation = type.GetTypeDefinition().Instantiation;
+                        }
+                        else
+                        {
+                            entityDefinitionInstantiation = ((MethodDesc)entityWithInstantiation).GetTypicalMethodDefinition().Instantiation;
+                        }
+
+                        GenericParameterDesc genericParam = (GenericParameterDesc)entityDefinitionInstantiation[iInstantiation];
+                        if (genericParam.Constraints.HasFlag(GenericConstraints.ReferenceTypeConstraint))
```

Comment:
This is shorter and avoids boxing:

```suggestion
                        if (genericParam.HasReferenceTypeConstraint)
```